File: app/src/main/java/com/thesunnahrevival/sunnahassistant/data/model/AppSettings.java
package com.thesunnahrevival.sunnahassistant.data.model;

import android.net.Uri;

import com.thesunnahrevival.sunnahassistant.utilities.SunnahAssistantUtil;

import java.util.Arrays;
import java.util.HashSet;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

@Entity(tableName = "app_settings")
public class AppSettings {

    @PrimaryKey(autoGenerate = true)
    private int id;
    private int method;
    private int asrCalculationMethod;
    private int month = 0;
    private int hijriOffSet = 0;
    private float latitude;
    private float longitude;
    private String formattedAddress;
    private boolean isLightMode = true;
    private boolean isFirstLaunch = true;
    private boolean showNextReminderNotification = true;
    private boolean showOnBoardingTutorial = true;
    private boolean isAutomatic;
    private boolean isDisplayHijriDate = true;
    private int savedSpinnerPosition = 0;
    private boolean isExpandedLayout = true;
    private Uri notificationToneUri;
    private boolean isVibrate = false;
    private int priority = 3;
    private int latitudeAdjustmentMethod = 3;
    private HashSet<String> categories = new HashSet<>( Arrays.asList(SunnahAssistantUtil.UNCATEGORIZED, SunnahAssistantUtil.SUNNAH, SunnahAssistantUtil.PRAYER, SunnahAssistantUtil.OTHER));

    public AppSettings(String formattedAddress, float latitude, float longitude, int method, int asrCalculationMethod, boolean isAutomatic) {
        this.formattedAddress = formattedAddress;
        this.latitude = latitude;
        this.longitude = longitude;
        this.method = method;
        this.asrCalculationMethod = asrCalculationMethod;
        this.isAutomatic = isAutomatic;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getFormattedAddress() {
        return formattedAddress;
    }

    public void setFormattedAddress(String formattedAddress) {
        this.formattedAddress = formattedAddress;
    }

    public float getLatitude() {
        return latitude;
    }

    public void setLatitude(float latitude) {
        this.latitude = latitude;
    }

    public float getLongitude() {
        return longitude;
    }

    public void setLongitude(float longitude) {
        this.longitude = longitude;
    }

    public int getMethod() {
        return method;
    }

    public void setMethod(int method) {
        this.method = method;
    }

    public int getAsrCalculationMethod() {
        return asrCalculationMethod;
    }

    public void setAsrCalculationMethod(int asrCalculationMethod) {
        this.asrCalculationMethod = asrCalculationMethod;
    }

    public boolean isAutomatic() {
        return isAutomatic;
    }

    public void setAutomatic(boolean automatic) {
        isAutomatic = automatic;
    }

    public int getMonth() {
        return month;
    }

    public void setMonth(int month) {
        this.month = month;
    }

    public int getHijriOffSet() {
        return hijriOffSet;
    }

    public void setHijriOffSet(int hijriOffSet) {
        this.hijriOffSet = hijriOffSet;
    }

    public boolean isLightMode() {
        return isLightMode;
    }

    public void setLightMode(boolean lightMode) {
        isLightMode = lightMode;
    }

    public boolean isFirstLaunch() {
        return isFirstLaunch;
    }

    public void setFirstLaunch(boolean firstLaunch) {
        isFirstLaunch = firstLaunch;
    }

    public boolean isShowNextReminderNotification() {
        return showNextReminderNotification;
    }

    public void setShowNextReminderNotification(boolean showNextReminderNotification) {
        this.showNextReminderNotification = showNextReminderNotification;
    }

    public boolean isShowOnBoardingTutorial() {
        return showOnBoardingTutorial;
    }

    public void setShowOnBoardingTutorial(boolean showOnBoardingTutorial) {
        this.showOnBoardingTutorial = showOnBoardingTutorial;
    }

    public boolean isDisplayHijriDate() {
        return isDisplayHijriDate;
    }

    public void setDisplayHijriDate(boolean displayHijriDate) {
        isDisplayHijriDate = displayHijriDate;
    }

    public int getSavedSpinnerPosition() {
        return savedSpinnerPosition;
    }

    public void setSavedSpinnerPosition(int savedSpinnerPosition) {
        this.savedSpinnerPosition = savedSpinnerPosition;
    }

    public HashSet<String> getCategories() {
        return categories;
    }

    public void setCategories(HashSet<String> categories) {
        this.categories = categories;
    }

    public boolean isExpandedLayout() {
        return isExpandedLayout;
    }

    public void setExpandedLayout(boolean expandedLayout) {
        isExpandedLayout = expandedLayout;
    }

    public Uri getNotificationToneUri() {
        return notificationToneUri;
    }

    public void setNotificationToneUri(Uri notificationToneUri) {
        this.notificationToneUri = notificationToneUri;
    }

    public boolean isVibrate() {
        return isVibrate;
    }

    public void setVibrate(boolean vibrate) {
        isVibrate = vibrate;
    }

    public int getPriority() {
        return priority;
    }

    public void setPriority(int priority) {
        this.priority = priority;
    }

    public int getLatitudeAdjustmentMethod() {
        return latitudeAdjustmentMethod;
    }

    public void setLatitudeAdjustmentMethod(int latitudeAdjustmentMethod) {
        this.latitudeAdjustmentMethod = latitudeAdjustmentMethod;
    }
}

File: app/src/main/java/com/thesunnahrevival/sunnahassistant/data/model/AppSettings.kt
package com.thesunnahrevival.sunnahassistant.data.model

import android.net.Uri
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.batoulapps.adhan.CalculationMethod
import com.batoulapps.adhan.Madhab
import com.thesunnahrevival.sunnahassistant.BuildConfig
import java.util.Date
import java.util.TreeSet

@Entity(tableName = "app_settings")
data class AppSettings(
    @PrimaryKey(autoGenerate = true) var id: Int = 1,
    var formattedAddress: String? = "",
    var latitude: Float = 0F,
    var longitude: Float = 0F,
    @ColumnInfo(name = "method") var calculationMethod: CalculationMethod = CalculationMethod.MUSLIM_WORLD_LEAGUE,
    var asrCalculationMethod: Madhab = Madhab.SHAFI,
    @ColumnInfo(name = "isAutomatic") var isAutomaticPrayerAlertsEnabled: Boolean = false,
    var enablePrayerTimeAlertsFor: BooleanArray = BooleanArray(5) { true },
    var prayerTimeOffsetsInMinutes: IntArray = IntArray(5) { 0 },
    var month: Int = 12,
    var hijriOffSet: Int = 0,
    var isLightMode: Boolean = true,
    var isFirstLaunch: Boolean = true,
    @ColumnInfo(name = "showNextReminderNotification") var showNextToDoNotification: Boolean = true,
    var showOnBoardingTutorial: Boolean = true,
    var isDisplayHijriDate: Boolean = true,
    var savedSpinnerPosition: Int = 0,
    var isExpandedLayout: Boolean = true,
    var notificationToneUri: Uri? = null,
    var isVibrate: Boolean = false,
    var priority: Int = 3,
    var latitudeAdjustmentMethod: Int = 2,
    var isShowHijriDateWidget: Boolean = true,
    @ColumnInfo(name = "isShowNextReminderWidget") var isShowNextToDoWidget: Boolean = true,
    var isAfterUpdate: Boolean = false,
    var appVersionCode: Int = BuildConfig.VERSION_CODE,
    var appVersion: String = BuildConfig.VERSION_NAME,
    var categories: TreeSet<String>? = null,
    var language: String = "en",
    var doNotDisturbMinutes: Int = 0,
    var useReliableAlarms: Boolean = true,
    var numberOfLaunches: Int = 0,
    var shareAnonymousUsageData: Boolean = true,
    @ColumnInfo(name = "generatePrayerRemindersAfter") var generatePrayerToDosAfter: Date = Date(),
    var includeHijriDateInCalendar: Boolean = true,
    @ColumnInfo(defaultValue = "0")
    var hideDownloadFilePrompt: Boolean = false
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as AppSettings

        if (id != other.id) return false
        if (formattedAddress != other.formattedAddress) return false
        if (latitude != other.latitude) return false
        if (longitude != other.longitude) return false
        if (calculationMethod != other.calculationMethod) return false
        if (asrCalculationMethod != other.asrCalculationMethod) return false
        if (isAutomaticPrayerAlertsEnabled != other.isAutomaticPrayerAlertsEnabled) return false
        if (!enablePrayerTimeAlertsFor.contentEquals(other.enablePrayerTimeAlertsFor)) return false
        if (!prayerTimeOffsetsInMinutes.contentEquals(other.prayerTimeOffsetsInMinutes)) return false
        if (month != other.month) return false
        if (hijriOffSet != other.hijriOffSet) return false
        if (isLightMode != other.isLightMode) return false
        if (isFirstLaunch != other.isFirstLaunch) return false
        if (showNextToDoNotification != other.showNextToDoNotification) return false
        if (showOnBoardingTutorial != other.showOnBoardingTutorial) return false
        if (isDisplayHijriDate != other.isDisplayHijriDate) return false
        if (savedSpinnerPosition != other.savedSpinnerPosition) return false
        if (isExpandedLayout != other.isExpandedLayout) return false
        if (notificationToneUri != other.notificationToneUri) return false
        if (isVibrate != other.isVibrate) return false
        if (priority != other.priority) return false
        if (latitudeAdjustmentMethod != other.latitudeAdjustmentMethod) return false
        if (isShowHijriDateWidget != other.isShowHijriDateWidget) return false
        if (isShowNextToDoWidget != other.isShowNextToDoWidget) return false
        if (isAfterUpdate != other.isAfterUpdate) return false
        if (appVersionCode != other.appVersionCode) return false
        if (appVersion != other.appVersion) return false
        if (categories != other.categories) return false
        if (language != other.language) return false
        if (doNotDisturbMinutes != other.doNotDisturbMinutes) return false
        if (useReliableAlarms != other.useReliableAlarms) return false
        if (numberOfLaunches != other.numberOfLaunches) return false
        if (shareAnonymousUsageData != other.shareAnonymousUsageData) return false
        if (generatePrayerToDosAfter != other.generatePrayerToDosAfter) return false
        if (includeHijriDateInCalendar != other.includeHijriDateInCalendar) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + (formattedAddress?.hashCode() ?: 0)
        result = 31 * result + latitude.hashCode()
        result = 31 * result + longitude.hashCode()
        result = 31 * result + calculationMethod.hashCode()
        result = 31 * result + asrCalculationMethod.hashCode()
        result = 31 * result + isAutomaticPrayerAlertsEnabled.hashCode()
        result = 31 * result + enablePrayerTimeAlertsFor.contentHashCode()
        result = 31 * result + prayerTimeOffsetsInMinutes.contentHashCode()
        result = 31 * result + month
        result = 31 * result + hijriOffSet
        result = 31 * result + isLightMode.hashCode()
        result = 31 * result + isFirstLaunch.hashCode()
        result = 31 * result + showNextToDoNotification.hashCode()
        result = 31 * result + showOnBoardingTutorial.hashCode()
        result = 31 * result + isDisplayHijriDate.hashCode()
        result = 31 * result + savedSpinnerPosition
        result = 31 * result + isExpandedLayout.hashCode()
        result = 31 * result + (notificationToneUri?.hashCode() ?: 0)
        result = 31 * result + isVibrate.hashCode()
        result = 31 * result + priority
        result = 31 * result + latitudeAdjustmentMethod
        result = 31 * result + isShowHijriDateWidget.hashCode()
        result = 31 * result + isShowNextToDoWidget.hashCode()
        result = 31 * result + isAfterUpdate.hashCode()
        result = 31 * result + appVersionCode
        result = 31 * result + appVersion.hashCode()
        result = 31 * result + (categories?.hashCode() ?: 0)
        result = 31 * result + language.hashCode()
        result = 31 * result + doNotDisturbMinutes
        result = 31 * result + useReliableAlarms.hashCode()
        result = 31 * result + numberOfLaunches
        result = 31 * result + shareAnonymousUsageData.hashCode()
        result = 31 * result + generatePrayerToDosAfter.hashCode()
        return result
    }
}